File: agent/nodes/classify.py
"""Classify node: extract question_type, buyer_persona, escalation_flags."""
from __future__ import annotations

import re
from typing import Any

from agent.llm import call_json, load_prompt
from agent.personas import count_keyword_hits, format_keywords_for_prompt
from agent.state import TicketState

VALID_QUESTION_TYPES = {
    "order_status",
    "engraving",
    "servicing",
    "strap_compatibility",
    "materials_safety",
    "product_general",
    "knowledge_gap",
}

VALID_BUYER_PERSONAS = {
    "health_conscious",
    "gifter",
    "enthusiast",
    "active",
    "sustainable",
}

VALID_ESCALATION_FLAGS = {
    "angry",
    "refund_overdue",
    "fulfilment_error",
    "corporate_bulk",
    "press_media",
    "high_liability",
    "order_id_mismatch",
    "older_model",
}


def _detect_order_id_mismatch(order_id_field: str, message_body: str) -> bool:
    """Catch the gotcha tickets where the field's order_id ≠ the body's order_id."""
    if not order_id_field:
        return False
    body_ids = re.findall(r"\bBLD-\d{4,}\b", message_body)
    if not body_ids:
        return False
    return any(bid != order_id_field for bid in body_ids)


def _clamp_confidence(value: Any) -> tuple[float, bool]:
    try:
        confidence = float(value)
    except (TypeError, ValueError):
        return 0.0, False
    if confidence < 0 or confidence > 1:
        return min(1.0, max(0.0, confidence)), False
    return confidence, True


def _validate_result(result: dict[str, Any]) -> tuple[str, str, list[str], float, list[str]]:
    errors = []

    question_type = result.get("question_type")
    if question_type not in VALID_QUESTION_TYPES:
        errors.append(f"invalid question_type: {question_type!r}")
        question_type = "product_general"

    buyer_persona = result.get("buyer_persona")
    if buyer_persona not in VALID_BUYER_PERSONAS:
        errors.append(f"invalid buyer_persona: {buyer_persona!r}")
        buyer_persona = "active"

    raw_flags = result.get("escalation_flags", [])
    if not isinstance(raw_flags, list):
        errors.append("escalation_flags was not a list")
        raw_flags = []

    flags = []
    for flag in raw_flags:
        if flag in VALID_ESCALATION_FLAGS:
            flags.append(flag)
        else:
            errors.append(f"invalid escalation flag: {flag!r}")

    confidence, confidence_valid = _clamp_confidence(result.get("confidence", 0.0))
    if not confidence_valid:
        errors.append(f"invalid confidence: {result.get('confidence')!r}")

    if errors and "classification_error" not in flags:
        flags.append("classification_error")

    return question_type, buyer_persona, flags, confidence, errors


def reconcile_persona(
    llm_persona: str,
    llm_confidence: float,
    keyword_hits: dict[str, int],
    _message_body: str,
) -> tuple[str, str]:
    """Combine the LLM's persona pick with deterministic trigger-keyword hits.

    The classify prompt now sees the trigger keywords in-context (so the LLM has
    the signal), but the LLM can still go off on subtle cases. The persona CSV
    gives us a deterministic second opinion. This function decides how to merge.

    Args:
        llm_persona: persona_id the LLM picked (already validated against the enum)
        llm_confidence: LLM's self-reported confidence in [0, 1]
        keyword_hits: {persona_id: number_of_trigger_keywords_matched_in_body}
        message_body: the original ticket text (in case the strategy wants context)

    Returns:
        (final_persona_id, reason)
            final_persona_id: the persona to use downstream
            reason: short audit-trail string (goes into state.notes)

    TODO ──────────────────────────────────────────────────────────────────────
    Pick a strategy and implement the body. Three options to consider — each has
    different failure modes. The default below is a NO-OP that trusts the LLM.

    Strategy A — "Hint only" (current default):
        Trust the LLM. The keywords are already in the prompt; the LLM has the
        signal. Just return what the LLM said. Simplest, no surprises.

    Strategy B — "Hard override on dominant keyword evidence":
        If exactly one persona has hits >= 2 AND it differs from llm_persona,
        override. Catches the case where the LLM ignored the keywords.
        Risk: a single "BPA" mention could override correct persona on a
        gifter ticket. Tune the threshold.

    Strategy C — "Confidence-weighted vote":
        Score each persona as (keyword_hits[p] * 0.4) + (1.0 if p == llm_persona else 0.0) * llm_confidence.
        Pick the argmax. Soft blend — both signals matter, neither dominates.

    Whichever you pick, set `reason` so the audit trail explains the decision.
    ───────────────────────────────────────────────────────────────────────────
    """
    # Strategy C — confidence-weighted vote
    # Each persona scores: (keyword_hits * 0.4) + (llm_confidence if it is the LLM's pick else 0)
    # Ties favour the LLM's pick for stability.
    scores: dict[str, float] = {}
    for persona_id, hits in keyword_hits.items():
        score = hits * 0.4
        if persona_id == llm_persona:
            score += llm_confidence
        scores[persona_id] = score

    best = max(scores, key=lambda p: (scores[p], p == llm_persona))
    if best == llm_persona:
        return llm_persona, f"vote kept LLM (top score {scores[best]:.2f})"
    return (
        best,
        f"vote override: {best} ({scores[best]:.2f}) > LLM={llm_persona} ({scores[llm_persona]:.2f})",
    )


def run(state: TicketState) -> TicketState:
    prompt = load_prompt("classify").format(
        order_id=state.get("order_id", "") or "(none provided)",
        channel=state.get("channel", "email"),
        subject=state.get("subject", ""),
        message_body=state.get("message_body", ""),
        persona_keywords=format_keywords_for_prompt(),
    )
    try:
        result = call_json(
            system="You are a precise customer service triage classifier. Output only valid JSON.",
            user=prompt,
            max_tokens=400,
        )
    except Exception as e:
        # Fail safe: route to human review
        state["question_type"] = "product_general"
        state["buyer_persona"] = "active"
        state["escalation_flags"] = ["classification_error"]
        state["classification_confidence"] = 0.0
        state.setdefault("notes", []).append(f"classify error: {e}")
        return state

    question_type, buyer_persona, flags, confidence, validation_errors = _validate_result(result)

    # Deterministic check — don't trust the LLM to spot order-ID mismatch reliably.
    if _detect_order_id_mismatch(state.get("order_id", ""), state.get("message_body", "")):
        if "order_id_mismatch" not in flags:
            flags.append("order_id_mismatch")

    # Reconcile the LLM's persona pick against trigger-keyword hits from the CSV.
    body_text = f"{state.get('subject', '')} {state.get('message_body', '')}"
    keyword_hits = count_keyword_hits(body_text)
    final_persona, reconcile_reason = reconcile_persona(
        buyer_persona,
        confidence,
        keyword_hits,
        body_text,
    )
    if final_persona != buyer_persona:
        state.setdefault("notes", []).append(
            f"persona reconcile: LLM={buyer_persona} → {final_persona} ({reconcile_reason})"
        )

    if validation_errors:
        state.setdefault("notes", []).append("classify validation: " + "; ".join(validation_errors))

    state["question_type"] = question_type
    state["buyer_persona"] = final_persona
    state["escalation_flags"] = flags
    state["classification_confidence"] = confidence
    state["persona_keyword_hits"] = keyword_hits  # for audit/debug + reasoning panel
    return state

File: web/lib/classify.ts
import { callJson, fillTemplate, loadPrompt } from "./llm";
import { countKeywordHits, formatKeywordsForPrompt } from "./personas";
import type { TicketState } from "./types";

function detectOrderIdMismatch(orderIdField: string, body: string): boolean {
  if (!orderIdField) return false;
  const bodyIds = body.match(/\bBLD-\d{4,}\b/g) ?? [];
  if (!bodyIds.length) return false;
  return bodyIds.some((bid) => bid !== orderIdField);
}

type ClassifyResult = {
  question_type?: string;
  buyer_persona?: string;
  escalation_flags?: unknown;
  confidence?: unknown;
};

const VALID_QUESTION_TYPES = new Set([
  "order_status",
  "engraving",
  "servicing",
  "strap_compatibility",
  "materials_safety",
  "product_general",
  "knowledge_gap",
]);

const VALID_BUYER_PERSONAS = new Set([
  "health_conscious",
  "gifter",
  "enthusiast",
  "active",
  "sustainable",
]);

const VALID_ESCALATION_FLAGS = new Set([
  "angry",
  "refund_overdue",
  "fulfilment_error",
  "corporate_bulk",
  "press_media",
  "high_liability",
  "order_id_mismatch",
  "older_model",
]);

function clampConfidence(value: unknown): [number, boolean] {
  const confidence = Number(value);
  if (!Number.isFinite(confidence)) return [0, false];
  if (confidence < 0 || confidence > 1) return [Math.min(1, Math.max(0, confidence)), false];
  return [confidence, true];
}

function validateResult(result: ClassifyResult): {
  questionType: string;
  buyerPersona: string;
  flags: string[];
  confidence: number;
  errors: string[];
} {
  const errors: string[] = [];

  let questionType = result.question_type;
  if (!questionType || !VALID_QUESTION_TYPES.has(questionType)) {
    errors.push(`invalid question_type: ${String(questionType)}`);
    questionType = "product_general";
  }

  let buyerPersona = result.buyer_persona;
  if (!buyerPersona || !VALID_BUYER_PERSONAS.has(buyerPersona)) {
    errors.push(`invalid buyer_persona: ${String(buyerPersona)}`);
    buyerPersona = "active";
  }

  const rawFlags = Array.isArray(result.escalation_flags) ? result.escalation_flags : [];
  if (result.escalation_flags !== undefined && !Array.isArray(result.escalation_flags)) {
    errors.push("escalation_flags was not an array");
  }

  const flags: string[] = [];
  for (const flag of rawFlags) {
    if (typeof flag === "string" && VALID_ESCALATION_FLAGS.has(flag)) {
      flags.push(flag);
    } else {
      errors.push(`invalid escalation flag: ${String(flag)}`);
    }
  }

  const [confidence, confidenceValid] = clampConfidence(result.confidence);
  if (!confidenceValid) errors.push(`invalid confidence: ${String(result.confidence)}`);

  if (errors.length && !flags.includes("classification_error")) flags.push("classification_error");

  return { questionType, buyerPersona, flags, confidence, errors };
}

/**
 * Combine the LLM's persona pick with deterministic trigger-keyword hits.
 *
 * The classify prompt now includes the trigger keywords (so the LLM has the
 * signal), but the LLM can still go off on subtle cases. The persona CSV gives
 * us a deterministic second opinion. This function decides how to merge.
 *
 * TODO ────────────────────────────────────────────────────────────────────────
 * Pick a strategy and implement the body. Three options worth considering —
 * each has different failure modes. The default below is a no-op that trusts
 * the LLM. Mirror your Python `reconcile_persona` choice here so the two
 * pipelines stay in agreement.
 *
 * A — "Hint only" (current default): trust the LLM since keywords are already
 *     in the prompt. No code needed.
 * B — "Hard override on dominant keyword evidence": if exactly one persona has
 *     hits >= 2 AND it differs from llmPersona, override.
 * C — "Confidence-weighted vote": score each persona as
 *     (keywordHits[p] * 0.4) + (p === llmPersona ? llmConfidence : 0); argmax.
 * ─────────────────────────────────────────────────────────────────────────────
 */
export function reconcilePersona(
  llmPersona: string,
  llmConfidence: number,
  keywordHits: Record<string, number>,
  _messageBody: string,
): { persona: string; reason: string } {
  // Strategy C — confidence-weighted vote
  // Each persona scores: (keyword_hits * 0.4) + (llmConfidence if it is the LLM's pick else 0)
  // Ties favour the LLM's pick for stability.
  const scores: Record<string, number> = {};
  for (const [pid, hits] of Object.entries(keywordHits)) {
    let s = hits * 0.4;
    if (pid === llmPersona) s += llmConfidence;
    scores[pid] = s;
  }

  let best = llmPersona;
  let bestScore = scores[llmPersona] ?? 0;
  for (const [pid, s] of Object.entries(scores)) {
    if (s > bestScore) {
      best = pid;
      bestScore = s;
    }
  }

  if (best === llmPersona) {
    return { persona: llmPersona, reason: `vote kept LLM (top score ${bestScore.toFixed(2)})` };
  }
  return {
    persona: best,
    reason: `vote override: ${best} (${bestScore.toFixed(2)}) > LLM=${llmPersona} (${(scores[llmPersona] ?? 0).toFixed(2)})`,
  };
}

export async function classify(state: TicketState): Promise<TicketState> {
  const prompt = fillTemplate(loadPrompt("classify"), {
    order_id: state.order_id || "(none provided)",
    channel: state.channel ?? "email",
    subject: state.subject ?? "",
    message_body: state.message_body ?? "",
    persona_keywords: formatKeywordsForPrompt(),
  });

  let result: ClassifyResult;
  try {
    result = await callJson<ClassifyResult>(
      "You are a precise customer service triage classifier. Output only valid JSON.",
      prompt,
      { maxTokens: 400 },
    );
  } catch (e) {
    return {
      ...state,
      question_type: "product_general",
      buyer_persona: "active",
      escalation_flags: ["classification_error"],
      classification_confidence: 0,
      notes: [...(state.notes ?? []), `classify error: ${(e as Error).message}`],
    };
  }

  const { questionType, buyerPersona, flags, confidence, errors } = validateResult(result);
  if (detectOrderIdMismatch(state.order_id ?? "", state.message_body ?? "")) {
    if (!flags.includes("order_id_mismatch")) flags.push("order_id_mismatch");
  }

  const bodyText = `${state.subject ?? ""} ${state.message_body ?? ""}`;
  const keywordHits = countKeywordHits(bodyText);
  const { persona: finalPersona, reason } = reconcilePersona(
    buyerPersona,
    confidence,
    keywordHits,
    bodyText,
  );

  const reconcileNote =
    finalPersona !== buyerPersona
      ? `persona reconcile: LLM=${buyerPersona} → ${finalPersona} (${reason})`
      : null;

  const newNotes = [
    ...(state.notes ?? []),
    ...(reconcileNote ? [reconcileNote] : []),
    ...(errors.length ? [`classify validation: ${errors.join("; ")}`] : []),
  ];

  return {
    ...state,
    question_type: questionType,
    buyer_persona: finalPersona,
    escalation_flags: flags,
    classification_confidence: confidence,
    persona_keyword_hits: keywordHits,
    notes: newNotes.length ? newNotes : state.notes,
  };
}
